pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,410 @@
+	})
+
+	pod := unstructuredFor(podGVK, podInstanceOne)
+	require.NoError(t, c.Create(ctx, pod), fmt.Sprintf("creating Pod %s", podInstanceOne))
+	t.Cleanup(func() {
+		assert.NoError(t, deleteResource(ctx, c, pod), fmt.Sprintf("deleting resource %s", podInstanceOne))
+	})
+
+	cfClient, ok := dataStore.(*cachemanager.FakeCfClient)
+	require.True(t, ok)
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	syncSourceTwo := aggregator.Key{Source: "source_b", ID: "ID_b"}
+
+	wg := &sync.WaitGroup{}
+
+	// simulate a churn-y concurrent access by swapping the gvks for the sync sources repeatedly
+	// and removing sync sources, all from different go routines.
+	for i := 1; i < 100; i++ {
+		wg.Add(3)
+		go func() {
+			defer wg.Done()
+
+			// add some jitter
+			time.Sleep(time.Duration(r.Intn(1000)) * time.Millisecond)
+
+			require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+			require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceTwo, []schema.GroupVersionKind{podGVK}))
+		}()
+		go func() {
+			defer wg.Done()
+
+			// add some jitter
+			time.Sleep(time.Duration(r.Intn(1000)) * time.Millisecond)
+
+			require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{podGVK}))
+			require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceTwo, []schema.GroupVersionKind{configMapGVK}))
+		}()
+		go func() {
+			defer wg.Done()
+
+			time.Sleep(time.Duration(r.Intn(1000)) * time.Millisecond)
+			require.NoError(t, cacheManager.RemoveSource(ctx, syncSourceTwo))
+
+			time.Sleep(time.Duration(r.Intn(1000)) * time.Millisecond)
+			require.NoError(t, cacheManager.RemoveSource(ctx, syncSourceOne))
+		}()
+	}
+
+	wg.Wait()
+
+	// final upsert for determinism
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceTwo, []schema.GroupVersionKind{podGVK}))
+
+	expected := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: nsedConfigInstanceOne}: nil,
+		{Gvk: configMapGVK, Key: nsedConfigInstanceTwo}: nil,
+		{Gvk: podGVK, Key: nsedPodInstanceOne}:          nil,
+	}
+
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+	// now assert that the gvkAggregator looks as expected
+	agg.IsPresent(configMapGVK)
+	gvks := agg.List(syncSourceOne)
+	require.Len(t, gvks, 1)
+	_, foundConfigMap := gvks[configMapGVK]
+	require.True(t, foundConfigMap)
+	gvks = agg.List(syncSourceTwo)
+	require.Len(t, gvks, 1)
+	_, foundPod := gvks[podGVK]
+	require.True(t, foundPod)
+
+	// now remove the sources
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		require.NoError(t, cacheManager.RemoveSource(ctx, syncSourceOne))
+	}()
+	go func() {
+		defer wg.Done()
+		require.NoError(t, cacheManager.RemoveSource(ctx, syncSourceTwo))
+	}()
+
+	wg.Wait()
```

Comment:
What is the value of multi-threading here?  Seems like a fancy way of writing:

```
require.NoError(t, cacheManager.RemoveSource(ctx, syncSourceOne))
require.NoError(t, cacheManager.RemoveSource(ctx, syncSourceTwo))
```

We have a much more substantial multi-threading behavior above with the 100X loop and jitter.

Reply:
removed